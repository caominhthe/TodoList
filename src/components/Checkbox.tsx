import React, { useState } from "react";
import { View, Text, TouchableOpacity, StyleSheet } from "react-native";
import { Palette, spacing } from "../theme";
import { Colors } from "../theme/color";

interface CheckBoxProps {
  onChange?: (checked: boolean) => void;
  value?: boolean;
}

const Checkbox: React.FC<CheckBoxProps> = ({ value = false, onChange }) => {
  const [isChecked, setIsChecked] = useState(value);

  const toggleCheckbox = () => {
    setIsChecked(!isChecked);
    if (onChange) {
      onChange(!isChecked);
    }
  };

  return (
    <TouchableOpacity onPress={toggleCheckbox} style={styles.container}>
      <View style={[styles.checkbox, isChecked ? styles.checked : null]}>
        {isChecked && <Text style={styles.checkmark}>âœ“</Text>}
      </View>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    alignItems: "center",
  },
  checkbox: {
    width: 20,
    height: 20,
    borderWidth: 2,
    borderColor: Colors.BackgroundDark,
    borderRadius: spacing.tiny,
    justifyContent: "center",
    alignItems: "center",
  },
  checked: {
    backgroundColor: Colors.BackgroundDark,
  },
  checkmark: {
    color: Palette.white,
    fontSize: 12,
  },
});

export default Checkbox;
