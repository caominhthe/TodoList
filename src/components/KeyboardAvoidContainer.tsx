import React, { useMemo } from "react";
import {
  View,
  ViewProps,
  StyleSheet,
  KeyboardAvoidingView,
  Platform,
} from "react-native";

import { useSafeAreaInsets } from "react-native-safe-area-context";
import { Palette } from "../theme";

interface KeyboardAvoidContainerProps extends ViewProps {
  withSafeKeyboard?: boolean;
  keyboardOffset?: number;
}

const KeyboardAvoidContainer: React.FC<KeyboardAvoidContainerProps> = ({
  style,
  children,
  withSafeKeyboard = false,
  keyboardOffset = 100,
  ...props
}) => {
  const insets = useSafeAreaInsets();

  const margin = useMemo(
    () => ({
      marginBottom: insets.bottom,
    }),
    [insets.bottom]
  );

  if (withSafeKeyboard) {
    return (
      <KeyboardAvoidingView
        style={styles.keyboard}
        enabled={Platform.OS === "ios"}
        behavior={"padding"}
        pointerEvents={"box-none"}
        keyboardVerticalOffset={-insets.bottom + keyboardOffset}
      >
        <View style={[styles.container, margin, style]} {...props}>
          {children}
        </View>
      </KeyboardAvoidingView>
    );
  }

  return (
    <View
      style={[styles.container, margin, style]}
      pointerEvents={"box-none"}
      {...props}
    >
      {children}
    </View>
  );
};

export default KeyboardAvoidContainer;
export type { KeyboardAvoidContainerProps };

const styles = StyleSheet.create({
  keyboard: {
    flex: 1,
    backgroundColor: Palette.white,
  },
  container: {
    flex: 1,
    backgroundColor: Palette.white,
  },
});
