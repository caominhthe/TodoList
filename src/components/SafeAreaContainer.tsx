import React, { useMemo } from "react";
import {
  View,
  ViewProps,
  StyleSheet,
  KeyboardAvoidingView,
  KeyboardAvoidingViewProps,
  Platform,
} from "react-native";

import { useSafeAreaInsets } from "react-native-safe-area-context";
import { Palette } from "../theme";

interface SafeAreaContainerProps extends ViewProps {
  withSafeKeyboard?: boolean;
  keyboardOffset?: number;
}

const SafeAreaContainer: React.FC<SafeAreaContainerProps> = ({
  style,
  children,
  withSafeKeyboard = false,
  keyboardOffset = 100,
  ...props
}) => {
  const insets = useSafeAreaInsets();

  const margin = useMemo(
    () => ({
      marginBottom: insets.bottom,
    }),
    [insets.bottom, insets.top]
  );

  if (withSafeKeyboard) {
    return (
      <KeyboardAvoidingView
        enabled={Platform.OS === "ios"}
        behavior={"padding"}
        style={styles.keyboard}
        pointerEvents={"box-none"}
        keyboardVerticalOffset={-insets.bottom + keyboardOffset}
      >
        <View
          style={[styles.container, margin, style]}
          pointerEvents={"box-none"}
          {...props}
        >
          {children}
        </View>
      </KeyboardAvoidingView>
    );
  }

  return (
    <View
      style={[styles.container, margin, style]}
      pointerEvents={"box-none"}
      {...props}
    >
      {children}
    </View>
  );
};

export default SafeAreaContainer;
export type { SafeAreaContainerProps };

const styles = StyleSheet.create({
  keyboard: {
    flex: 1,
    backgroundColor: Palette.white,
  },
  container: {
    flex: 1,
    backgroundColor: Palette.white,
  },
});
