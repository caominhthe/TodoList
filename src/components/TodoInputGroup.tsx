import React from "react";
import {
  Dimensions,
  StyleSheet,
  TextInput,
  TouchableOpacity,
  View,
} from "react-native";
import { spacing, typography } from "../theme";
import { Colors } from "../theme/color";

interface TodoInputGroup {
  value?: string;
  onChangeText?: (text: string) => void;
  onPressButton?: () => void;
  onBlur?: () => void;
  button: React.ReactNode | any;
  inputRef?: React.Ref<TextInput>;
}

const TodoInputGroup: React.FC<TodoInputGroup> = ({
  value,
  onChangeText,
  onPressButton,
  onBlur,
  button,
  inputRef,
}) => {
  return (
    <View style={styles.textInputContainer}>
      <TextInput
        ref={inputRef}
        placeholder="Add new todo here"
        onChangeText={onChangeText}
        onBlur={onBlur}
        style={styles.textInput}
        multiline
        value={value}
      />
      <TouchableOpacity testID="submit-button" onPress={onPressButton}>
        {button}
      </TouchableOpacity>
    </View>
  );
};

export default TodoInputGroup;

const { width: windowWidth } = Dimensions.get("window");

const styles = StyleSheet.create({
  textInputContainer: {
    flexDirection: "row",
    alignItems: "center",
    paddingHorizontal: spacing.medium,
    paddingBottom: spacing.tiny,
    width: windowWidth,
  },
  textInput: {
    flex: 1,
    fontSize: typography.fontSize.large,
    lineHeight: 30,
    borderWidth: 1,
    borderColor: Colors.BackgroundDark,
    borderRadius: spacing.small,
    paddingHorizontal: spacing.small,
    paddingVertical: spacing.small,
    marginRight: spacing.small,
  },
});
