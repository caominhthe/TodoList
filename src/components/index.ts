export { default as TodoItem } from "./TodoItem";
export { default as TodoInputGroup } from "./TodoInputGroup";
export { default as KeyboardAvoidContainer } from "./KeyboardAvoidContainer";
