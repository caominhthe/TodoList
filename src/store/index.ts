import { applyMiddleware, compose, createStore } from "redux";

import persistCombineReducers from "redux-persist/es/persistCombineReducers";
import persistStore from "redux-persist/es/persistStore";
import rootReducer from "./reducer";
import thunk from "redux-thunk";
import { reduxStorage } from "../reducers/reduxStorage";

declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__?: typeof compose;
  }
}

const persistConfig = {
  key: "root",
  storage: reduxStorage,
  blacklist: ["todo"],
};

const persistedReducer = persistCombineReducers(persistConfig, rootReducer);

export const store = createStore(persistedReducer, applyMiddleware(thunk));

export const persistor = persistStore(store);
