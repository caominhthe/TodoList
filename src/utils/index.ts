import { Todo } from "../screens";

export function createTodoItem(content: string) {
  const todo: Todo = {
    id: String(Date.now()),
    title: content,
    createdAt: Date.now(),
    isDone: false,
  };
  return todo;
}
