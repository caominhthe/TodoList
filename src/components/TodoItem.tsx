import {
  StyleSheet,
  TouchableOpacity,
  View,
  Text,
  TextStyle,
  Pressable,
} from "react-native";
import React, { useMemo } from "react";
import { useCallback } from "react";
import { spacing, typography } from "../theme";
import Animated, { SlideInDown, SlideOutLeft } from "react-native-reanimated";
import { Todo } from "../screens";
import Checkbox from "./Checkbox";
import { Colors } from "../theme/color";

interface TodoItemProps {
  todo: Todo;
  isSelected: boolean;
  onDelete: (todo: Todo) => void;
  onPressCheckBox: (todo: Todo) => void;
  onPress: (todo: Todo) => void;
}

const TodoItem: React.FC<TodoItemProps> = ({
  todo,
  onPress,
  onDelete,
  onPressCheckBox,
  isSelected,
}) => {
  const onPressTodo = useCallback(() => {
    onPress?.(todo);
  }, [onPress, todo]);

  const onPressDelete = useCallback(() => {
    onDelete?.(todo);
  }, [onDelete, todo]);

  const textStyle = useMemo((): TextStyle => {
    return todo.isDone ? { textDecorationLine: "line-through" } : {};
  }, [todo.isDone]);

  return (
    <Pressable onPress={onPressTodo}>
      <Animated.View
        style={[styles.container, isSelected && styles.isSelected]}
        key={todo.id}
        entering={SlideInDown.duration(200)}
        exiting={SlideOutLeft.duration(200)}
      >
        <Checkbox onChange={() => onPressCheckBox(todo)} value={todo.isDone} />
        <View testID="todo-item" style={styles.titleButton}>
          <Text testID="item-title" style={[styles.title, textStyle]}>
            {todo.title}
          </Text>
        </View>
        <TouchableOpacity testID="delete-button" onPress={onPressDelete}>
          <Text style={styles.delete}>Delete</Text>
        </TouchableOpacity>
      </Animated.View>
    </Pressable>
  );
};

export default TodoItem;

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    alignItems: "center",
    minHeight: 60,
    paddingVertical: spacing.small,
    borderWidth: 1,
    borderColor: "grey",
    borderRadius: spacing.small,
    paddingHorizontal: spacing.large,
    paddingTop: spacing.small,
    marginTop: spacing.small,
    marginHorizontal: spacing.large,
  },
  delete: {
    color: "blue",
    fontSize: typography.fontSize.medium,
  },
  rightIcon: {
    marginRight: spacing.medium,
    marginLeft: "auto",
  },
  leftIcon: {
    marginLeft: spacing.medium,
  },
  title: {
    paddingHorizontal: spacing.medium,
    fontSize: typography.fontSize.medium,
    fontWeight: "bold",
    flexWrap: "wrap",
  },
  titleButton: {
    flex: 1,
  },
  isSelected: {
    backgroundColor: Colors.Background,
  },
});
