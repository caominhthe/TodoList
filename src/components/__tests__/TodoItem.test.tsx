import { TodoItem } from "@root/src/components";

import renderer from "react-test-renderer";
import { render, fireEvent } from "@testing-library/react-native";
import { createTodoItem } from "@root/src/utils";

describe("TodoItem", () => {
  const randomTodo = createTodoItem("test");
  const onPressEvent = jest.fn();
  const onDeleteEvent = jest.fn();
  const onPressCheckBox = jest.fn();

  it("should render correctly", async () => {
    const todo = renderer.create(
      <TodoItem
        todo={randomTodo}
        onPress={onPressEvent}
        onDelete={onDeleteEvent}
        isSelected={false}
        onPressCheckBox={onPressCheckBox}
      />
    );
    const todoJSON = todo.toJSON();
    expect(todoJSON).toMatchSnapshot();
  });

  it("should call action on Press", async () => {
    const { getByTestId } = render(
      <TodoItem
        todo={randomTodo}
        onPress={onPressEvent}
        onDelete={onDeleteEvent}
        isSelected={false}
        onPressCheckBox={onPressCheckBox}
      />
    );
    expect(getByTestId("item-title").props.children).toBe(randomTodo.title);

    const deleteButton = getByTestId("delete-button");
    fireEvent.press(deleteButton);

    expect(onDeleteEvent.mock.calls.length).toBe(1);

    const item = getByTestId("todo-item");
    fireEvent.press(item);
    expect(onPressEvent.mock.calls.length).toBe(1);
  });
});
