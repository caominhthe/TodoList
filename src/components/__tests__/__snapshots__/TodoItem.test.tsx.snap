// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem should render correctly 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "borderColor": "grey",
        "borderRadius": 8,
        "borderWidth": 1,
        "flexDirection": "row",
        "marginHorizontal": 24,
        "marginTop": 8,
        "minHeight": 40,
        "paddingHorizontal": 24,
        "paddingTop": 8,
        "paddingVertical": 8,
      },
      false,
    ]
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "flex": 1,
        "opacity": 1,
      }
    }
    testID="todo-item"
  >
    <Text
      style={
        {
          "flexWrap": "wrap",
          "fontSize": 16,
          "fontWeight": "bold",
          "paddingHorizontal": 12,
        }
      }
      testID="item-title"
    >
      test
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="delete-button"
  >
    <Text
      style={
        {
          "color": "blue",
          "fontSize": 16,
        }
      }
    >
      Delete
    </Text>
  </View>
</View>
`;
